src/hooks/useUsersGuildPins.ts

```diff
@@ -8,7 +8,101 @@ import {
   GUILD_PIN_CONTRACTS,
   GuildPinsSupportedChain,
 } from "utils/guildCheckout/constants"
-import { createPublicClient, http } from "viem"
+import { PublicClient, createPublicClient, http } from "viem"
+
+const getUsersGuildPinIdsOnChain = async (
+  balance: bigint,
+  chain: GuildPinsSupportedChain,
+  address: `0x${string}`,
+  client: PublicClient
+) => {
+  const contracts = Array.from({ length: Number(balance) }, (_, i) => ({
+    abi: GUILD_PIN_CONTRACTS[chain].abi,
+    address: GUILD_PIN_CONTRACTS[chain].address,
+    functionName: "tokenOfOwnerByIndex",
+    args: [address, BigInt(i)],
+  }))
+
+  const results =
+    contracts.length > 0
+      ? await client.multicall({
+          contracts: contracts,
+        })
+      : []
+
+  const errors: Error[] = results
+    .filter((result) => result.status === "failure")
+    .map((result) => result.error)
+
+  const pinIds = results
+    .filter((result) => result.status === "success")
+    .map((result) => result.result as bigint)
+
+  return { pinIds, errors }
+}
+
+type TokenInfo = {
+  chainId: Chains
+  tokenId: number
+  tokenUri: string
+}
+
+const getPinTokenURIsForPinIds = async (
+  pinIds: bigint[],
+  chain: GuildPinsSupportedChain,
+  client: PublicClient
+) => {
+  const contractCalls = pinIds.map((tokenId) => ({
+    abi: GUILD_PIN_CONTRACTS[chain].abi,
+    address: GUILD_PIN_CONTRACTS[chain].address,
+    functionName: "tokenURI",
+    args: [tokenId],
+  }))
+
+  const results =
+    contractCalls.length > 0
+      ? await client.multicall({
+          contracts: contractCalls,
+        })
+      : []
+
+  const tokenInfo: TokenInfo[] = results.map((result, index) => ({
+    chainId: Chains[chain],
+    tokenId: Number(pinIds[index]),
+    tokenUri: result.status === "success" ? (result.result as string) : null,
+  }))
+
+  const errors: Error[] = results
+    .filter((res) => res.status === "failure")
+    .map((res) => res.error)
+
+  return { tokenInfo, errors }
+}
+
+type TokenWithMetadata = {
+  chainId: Chains
+  tokenId: number
+  image: string
```

Comment:
I think image is already defined in `GuildPinMetadata` ;)

Reply:
Nice, do you have a tip for noticing these typing opportunities?

Reply:
I'll be honest, I noticed by chance :D
But, I guess you'll just get to know the codebase better over time and you'll notice these too ;)